src/core/geometry/qgscurvepolygon.cpp

```diff
@@ -435,6 +442,52 @@ QgsAbstractGeometry *QgsCurvePolygon::boundary() const
   }
 }
 
+QgsCurvePolygon *QgsCurvePolygon::asGridified( double hSpacing, double vSpacing, double dSpacing, double mSpacing,
+    double tolerance, SegmentationToleranceType toleranceType ) const
+{
+  if ( !mExteriorRing )
+    return nullptr;
+
+  QgsPolygonV2 *polygon;
+  if ( QgsWkbTypes::flatType( mWkbType ) == QgsWkbTypes::Triangle ||  QgsWkbTypes::flatType( mWkbType ) == QgsWkbTypes::Polygon )
+  {
+    polygon = static_cast< QgsPolygonV2 const *>( this )->newSameGeometry();
+  }
+  else
+  {
+    polygon = new QgsPolygonV2();
+    polygon->mWkbType = QgsWkbTypes::zmType( QgsWkbTypes::Polygon, QgsWkbTypes::hasZ( mWkbType ), QgsWkbTypes::hasM( mWkbType ) );
+  }
+
+  // exterior ring
+  QgsCurve *exterior = mExteriorRing->asGridified( hSpacing, vSpacing, dSpacing, mSpacing, tolerance, toleranceType );
+
+  if ( !exterior )
+  {
+    delete polygon;
+    return nullptr;
+  }
+
+  polygon->mExteriorRing = exterior;
```

Comment:
Better to call `polygon->setExteriorRing( exterior )` here. setExteriorRing has extra logic which may be required.

Reply:
The extra logic is the following: delete old exterior ring, change the type (which we already done manually) and clear cache (which we don't have). In case of `QgsPolygonV2` it also looks if segmentation is needed (it never is).